configurable user session TTL

akostadinov opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Please remember to ALWAYS open an issue before starting to work on your pull request. Please take the time to validate your intentions for the pull request with the project maintainers before spending the time to work on it, so your time does not go to waste. 
2. If this is your first time, please make sure you've gone through the Contribution guide.
3. If the PR is unfinished, add a `[WIP]` at the start of the PR title. You can remove it when it's ready to be reviewed.
-->

**What this PR does / why we need it**:

Add support to configure user session time to live [THREESCALE-693](https://issues.redhat.com/browse/THREESCALE-693)

**Which issue(s) this PR fixes** 

fixes THREESCALE-693

**Verification steps** 

1. set `user_sesion_ttl` in `settings.yml` to something small, like 1 minute `PT1M`
2. run porta 
3. open some porta provider console
4. login
5. check that you can navigate between pages
6. wait without activity for more than 1 minute
7. try to navigate to another page
8. you should be asked to login again

**Special notes for your reviewer**:

I couldn't figure out what other tests I can add. Existing tests already vefiry default TTL of 2 weeks

Still under investigation what needs to be added to OpenShift operator. In any case it should work as before OOB so operator changes can be applied any time after this is merged.

Another question is where should we document this?

guicassolato commented on app/models/user_session.rb:
```diff
@@ -1,6 +1,15 @@
+  TTL = case Rails.configuration.three_scale.user_sesion_ttl
+        when nil, ""
+          2.weeks
+        when Numeric, /^\d*$/
```
```suggestion
        when Numeric, /^\d+$/
```

akostadinov replied:
Good catch, thanks!

guicassolato commented on app/models/user_session.rb:
```diff
@@ -1,6 +1,15 @@
 class UserSession < ApplicationRecord
 
-  TTL = 2.weeks.freeze
+  TTL = case Rails.configuration.three_scale.user_sesion_ttl
+        when nil, ""
+          2.weeks
+        when Numeric, /^\d+$/
+          # session duration specified in seconds
+          ActiveSupport::Duration.parse("PT#{Rails.configuration.three_scale.user_sesion_ttl}S")
+        else
+          # treat as ISO 8601 duration
+          ActiveSupport::Duration.parse(Rails.configuration.three_scale.user_sesion_ttl)
+        end.freeze
```
With a a bit of DRY maybe:

```suggestion
  user_sesion_ttl = Rails.configuration.three_scale.user_sesion_ttl.presence
  TTL = begin
          case user_sesion_ttl
          when Numeric, /^\d+$/ # session duration specified in seconds
            ActiveSupport::Duration.parse("PT#{user_sesion_ttl}S")
          else
            ActiveSupport::Duration.parse(user_sesion_ttl)
          end
        rescue ActiveSupport::Duration::ISO8601Parser::ParsingError, TypeError
          # do we care about logging?
          2.weeks
        end.freeze
```

akostadinov replied:
I usually prefer to have things fail instead of silently ignoring user specified configuration. I'm also going to remove ISO8601 parsing per other comments. Do you have a strong preference to apply default configuration in case of a bad user-supplied value?

guicassolato replied:
This is not a critical feature, therefore, in case of bad user input, I'd be happy enough with falling back to a safe default plus a log message. However, this is no strong preference of mine, no. You should follow your heart 🙂

akostadinov replied:
because there are security concerns about session ttl, I'd rather fail. Please check what kind of exception I'm returning, I was not sure I chose the correct one.

hallelujah commented on config/examples/settings.yml:
```diff
@@ -14,6 +14,8 @@ base: &default
   recaptcha_public_key: <%= ENV.fetch('RECAPTCHA_PUBLIC_KEY', 'YOUR_RECAPTCHA_PUBLIC_KEY') %>
   recaptcha_private_key: <%= ENV['RECAPTCHA_PRIVATE_KEY'] %>
+  # configure user session timeout in seconds or ISO 8601 duration
+  user_sesion_ttl: <%= ENV.fetch('USER_SESSION_TTL', "P14D") %>
```
```suggestion
  user_sesion_ttl: <%= ENV.fetch('USER_SESSION_TTL',  14.days.seconds ) %>
```

hallelujah commented on openshift/system/config/settings.yml:
```diff
@@ -14,6 +14,8 @@ base: &default
   recaptcha_public_key: <%= ENV.fetch('RECAPTCHA_PUBLIC_KEY', 'YOUR_RECAPTCHA_PUBLIC_KEY') %>
   recaptcha_private_key: <%= ENV['RECAPTCHA_PRIVATE_KEY'] %>
+  # configure user session timeout in seconds or ISO 8601 duration
+  user_sesion_ttl: <%= ENV.fetch('USER_SESSION_TTL', "P14D") %>
```
```suggestion
  user_sesion_ttl: <%= ENV.fetch('USER_SESSION_TTL',  14.days.seconds ) %>
```

akostadinov commented on app/models/user_session.rb:
```diff
@@ -1,6 +1,15 @@
 class UserSession < ApplicationRecord
 
-  TTL = 2.weeks.freeze
+  TTL = case Rails.configuration.three_scale.user_sesion_ttl
```
I also don't find the constant choice very flexible. This is the least disruptive way to implement the feature and it can change whenever we decide to allow configurable sessions per tenant and/or portal. Will keep constant if you don't mind.

will change to only seconds

akostadinov commented on config/docker/settings.yml:
```diff
@@ -14,6 +14,8 @@ base: &default
   recaptcha_public_key: <%= ENV.fetch('RECAPTCHA_PUBLIC_KEY', 'YOUR_RECAPTCHA_PUBLIC_KEY') %>
   recaptcha_private_key: <%= ENV['RECAPTCHA_PRIVATE_KEY'] %>
+  # configure user session timeout in seconds or ISO 8601 duration
+  user_sesion_ttl: <%= ENV.fetch('USER_SESSION_TTL', "P14D") %>
```
Makes perfect sense, thank you.

akostadinov commented on app/models/user_session.rb:
```diff
@@ -1,6 +1,13 @@
 class UserSession < ApplicationRecord
 
-  TTL = 2.weeks.freeze
+  TTL = case Rails.configuration.three_scale.user_sesion_ttl
```
@hallelujah , left it as constant for least destructive way for the change and leave refactoring for the time we decide to make configurable on a finer grain. Let me know if you are ok with that.

hallelujah replied:
I really prefer not to bother with those if cases

Those are settings, we are not supposed to guard against all things a user could do. Keep it simple

akostadinov replied:
@hallelujah , what change exactly do you suggest? To remove the `else` block or remove all conditions? If it is the latter, it will make IMO more complicated usage as all callers will need to do these checks instead of having them here. Or otherwise we'll have to change this to a method or something. Please let me know.

hallelujah replied:
yes just remove the check, this is not needed really to my opinion

```suggestion
  TTL = Rails.configuration.three_scale.user_sesion_ttl
```

guicassolato replied:
The 3scale operator usually prefers implementation where we prepare for env vars to be set with empty strings.

hallelujah replied:
This is different. While we have to do that more often, better have it in one place for all those settings. And not have sparse code that takes care of configuration values in multiple files. In fact I am really thinking of getting rid of such constants defined dynamically and have them gathered in one single `AppSettings` instance

Fair enough, I think this is OK for now but I'd suggest to make another PR to that, below a proposed approach:


```ruby
# lib/app_settings.rb
class AppSettings
   include Singleton

  attr_reader :config

  def initialize
     # Or just include ActiveSupport::Configurable ... everything works
    @config = ActiveSupport::Configurable::Configuration.new
    super
  end

  def self.configure(key, &block)
    instance.configure(key, &block)
  end

  def configure(key, &processor)
    config[key] = processor.call
  end

  def get(key)
    config.fetch key
  end

  def self.get(key)
    instance.get(key)
  end
end
```

Then you can use it in an initializer

```ruby
# config/initializers/configuration.rb
require 'app_settings'

AppSettings.configure('user_session.ttl') do
 ENV['USER_SESSION_TTL'].presence? || 2.weeks
end
```

That way we could probably get rid of the logic in the `config/*.yml` files and have a single way to require those configuration:

```ruby
 def remember_me   
   remember_me_for AppSettings.get('user_session.ttl')
end
```

guicassolato replied:
I like this idea, though I'd think of a path to make `AppSettings` a proper replacement for `Rails.settings.three_scale` in the future, rather than yet another settings system.

hallelujah replied:
Yes exactly what I was thinking. Juggling with different files inside of the config/application.rb might not be the proper way to do that.

I propose would be:

1. YAML config files should be free from logic, it is hard to override them with https://github.com/3scale/3scale-amp-openshift-templates/blob/baed04acb897f5d3d25ab5fddb2fcc3778c22a0d/amp/amp.yml#L2176-L2177
2. cleanup this mess https://github.com/3scale/porta/blob/61a168335181a09d1671da684a49f23078fe933d/config/application.rb#L143-L210
3. Try to put code logic where it belongs to, meaning in a class or initializer
  * not int the config/application.rb
4. Ensure the values are well formatted

guicassolato reviewed (dismissed):
Good job. Let's not forget the rest of this change has to be done in https://github.com/3scale/3scale-operator.

hallelujah reviewed (changes requested):
I think having a constant that just refers to the config is enough, Do not guard against the user input as it is done at boot time

hallelujah commented:
Looks good, can you fix codeclimate errors?

hallelujah commented on app/models/user_session.rb:
```diff
@@ -56,6 +54,19 @@ def user_agent=(value)
     self[:user_agent] = value
   end
 
+  private_class_method def self.user_session_ttl_parse(duration)
```
By the way you should just put this method one block below and you do not have to use `private_class_method`

akostadinov replied:
I moved it. With `private_class_method` I wanted to make clear it is an internal utility method not for reuse. And I have to leave it, otherwise robocop gives a warning

```
private (on line 57) does not make singleton methods private. Use private_class_method or private inside a class << self block instead.
```

Moving to a utility class IMO better be done once your suggestion above for a new configuration class is implemented but let me know if you want to see it elsewhere.

Please review again. Hope no other validator will find an issue with it now.

hallelujah replied:
I was thinking of that yes, sorry I did not make myself clear enough

private

class << self
  def user_session_ttl_parse(duration)
    # ...
  end
end

hallelujah commented on app/models/user_session.rb:
```diff
@@ -60,4 +58,18 @@ def user_agent=(value)
+
+  private_class_method def self.user_session_ttl_parse(duration)
+    case duration
+    when NilClass, ""
```
you can even do 
```suggestion
    when duration.blank?
```

hallelujah commented on app/models/user_session.rb:
```diff
@@ -60,4 +58,18 @@ def user_agent=(value)
+      ActiveSupport::Duration.seconds(duration.to_i)
+    else
+      raise ArgumentError, "user session TTL should be specified as a number of seconds"
+    end.freeze
```
Hmmm do you want to have the object returned by the method frozen? I would not expect that.

Another thing instead of the `case .. when`

```ruby
def user_session_ttl_parse(duration)
  return 2.weeks if duration.blank?

  ActiveSupport::Duration.seconds(Integer(duration))
rescue ArgumentError
  raise ApplicationConfigurationError, "user session TTL should be specified as a number of seconds"
end
```
Creating a tagged class to know exactly where it belongs to

hallelujah reviewed (changes requested):
I think the method should not freeze the result

akostadinov commented:
Moved freezing, added a test for it, renamed and refactored utility method, moved it inside a `class << self`

I'm concerned about moving method to `class <<self` block because it is public now and other public class methods are already defined as `def self.mymethod` within the same file so this is kind of inconsistent. Or did you mean to add `private` keyword inside the block? Sorry I'm a little confused about the intention of some of the requested changes.

I also didn't define `DEFAULT_TTL` constant to avoid extending public API with something that may not be intended to be used from outside the class. Not sure whether defining that was important to you. I can define it if you want.

hallelujah commented:
About the public method it is fine, I think it is clear now that it is a utility method but that is fine to leave it there for now. Once we do the refactoring of all the settings, we can extract it.
Also testing a private method is not really testable are they?
